src/include/sof/audio/audio_stream.h

```diff
@@ -15,6 +15,10 @@
 #define __SOF_AUDIO_AUDIO_STREAM_H__
 
 #include <sof/audio/format.h>
+#include <sof/audio/sink_api.h>
+#include <sof/audio/sink_api_implementation.h>
+#include <sof/audio/source_api.h>
+#include <sof/audio/source_api_implementation.h>
```

Comment:
again both implementers and users will get these headers. So, yes, they are in effect public.

Reply:
here everything is "private"
What do you think about:

public:
`  struct * source;  -  will never be defined, intended to use as a pointer only`

implementation:
```
  struct source_private {
   int a, b ,c; 
   }
 
```
and in .c fie:
```

  define get_source_priv(_src) (struct source_private *)_src 
  
  int source_something(struct *source _source) 
  {
      struct source_private *source = get_source_priv(_source);
    
      source->a =.... 
  }
```

Reply:
I don't see what this brings - implementers, users, etc. anyway have to know the size to allocate / include in their structures, and if they use `struct source_private` for that - what have we won?

Reply:
you're right
Looks like there's no way to block access at compiler level without mallocs

Reply:
@marcinszkudlinski @lyakh  I guess this depends on how much encapsulation we want. I don't think blocking access is a huge deal, but I if we want to have sink/source implementations that have a vastly more complex implementation, it's very likely they require some of their own state. So if we need this, using a private pointer plus a type-specific init, would seem like a good, well-known solution. Should be familiar to C++ folks as well (as the d-pointer pattern -> https://en.wikipedia.org/wiki/Opaque_pointer ). 
Simple types (like audio_stream.h) could just leave the private pointer unused and no heap/malloc would be needed (in our current uses).

We can surely do this incrementally. We only have one implmentation now, so there's quite the risk of overengineering here without actual users and as long as client code just accesses the public functions, were are free to modify the implementation in follow-up PRs.